--- a/net/minecraft/world/entity/projectile/Projectile.java
+++ b/net/minecraft/world/entity/projectile/Projectile.java
@@ -88,7 +_,7 @@
 
     @Nullable
     public Entity getOwnerRaw() { // Folia - region threading
-        ca.spottedleaf.moonrise.common.util.TickThread.ensureTickThread(this, "Cannot update owner state asynchronously"); // Folia - region threading
+        if (!ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(this)) return null; // okocraft - Fix crash when loading thrown enderpearl
         if (this.cachedOwner != null && !this.cachedOwner.isPurged()) { // Folia - region threading
             this.refreshProjectileSource(false); // Paper - Refresh ProjectileSource for projectiles
             return this.cachedOwner.getHandleRaw(); // Folia - region threading
