--- a/net/minecraft/stats/ServerRecipeBook.java
+++ b/net/minecraft/stats/ServerRecipeBook.java
@@ -110,6 +_,7 @@
     ) {
         for (ResourceKey<Recipe<?>> recipe : recipes) {
             if (!validator.test(recipe)) {
+                if (true) continue; // okocraft - Disable recipe warnings
                 LOGGER.error("Tried to load unrecognized recipe: {} removed now.", recipe);
             } else {
                 recipeAddingMethod.accept(recipe);
