--- a/src/main/java/org/spigotmc/event/player/PlayerSpawnLocationEvent.java
+++ b/src/main/java/org/spigotmc/event/player/PlayerSpawnLocationEvent.java
@@ -37,6 +_,7 @@
      * @param location the spawn location
      */
     public void setSpawnLocation(@NotNull Location location) {
+        org.bukkit.Bukkit.getLogger().log(java.util.logging.Level.WARNING, "PlayerSpawnLocationEvent#setSpawnLocation has no effect in this server!", new Throwable()); // okocraft - Notify that PlayerSpawnLocationEvent#setSpawnLocation has no effect
         this.spawnLocation = location.clone();
     }
 
