--- a/net/minecraft/server/PlayerAdvancements.java
+++ b/net/minecraft/server/PlayerAdvancements.java
@@ -51,10 +_,10 @@
     private final PlayerList playerList;
     private final Path playerSavePath;
     private AdvancementTree tree;
-    private final Map<AdvancementHolder, AdvancementProgress> progress = new LinkedHashMap<>();
-    private final Set<AdvancementHolder> visible = new HashSet<>();
-    private final Set<AdvancementHolder> progressChanged = new HashSet<>();
-    private final Set<AdvancementNode> rootsToUpdate = new HashSet<>();
+    private final Map<AdvancementHolder, AdvancementProgress> progress = java.util.Collections.synchronizedMap(new LinkedHashMap<>()); // okocraft - region threading
+    private final Set<AdvancementHolder> visible = java.util.Collections.synchronizedSet(new HashSet<>()); // okocraft - region threading
+    private final Set<AdvancementHolder> progressChanged = java.util.Collections.synchronizedSet(new HashSet<>()); // okocraft - region threading
+    private final Set<AdvancementNode> rootsToUpdate = java.util.Collections.synchronizedSet(new HashSet<>()); // okocraft - region threading
     private ServerPlayer player;
     @Nullable
     private AdvancementHolder lastSelectedTab;
@@ -146,7 +_,7 @@
         data.forEach((path, progress) -> {
             AdvancementHolder advancementHolder = advancementManager.get(path);
             if (advancementHolder == null) {
-                if (!path.getNamespace().equals(ResourceLocation.DEFAULT_NAMESPACE)) return; // CraftBukkit
+                if (true || !path.getNamespace().equals(ResourceLocation.DEFAULT_NAMESPACE)) return; // CraftBukkit // okocraft - Disable advancement warning
                 LOGGER.warn("Ignored advancement '{}' in progress file {} - it doesn't exist anymore?", path, this.playerSavePath);
             } else {
                 this.startProgress(advancementHolder, progress);
