--- a/net/minecraft/world/entity/ai/behavior/YieldJobSite.java
+++ b/net/minecraft/world/entity/ai/behavior/YieldJobSite.java
@@ -32,6 +_,7 @@
                             return false;
                         } else if (villager.getVillagerData().getProfession() != VillagerProfession.NONE) {
                             return false;
+                        } else if (!ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(level, villager.position())) { return false; // okocraft - region threading
                         } else {
                             BlockPos blockPos = instance.<GlobalPos>get(potentialJobSite).pos();
                             Optional<Holder<PoiType>> type = level.getPoiManager().getType(blockPos);
