--- a/net/minecraft/world/entity/ai/village/VillageSiege.java
+++ b/net/minecraft/world/entity/ai/village/VillageSiege.java
@@ -100,7 +_,7 @@
     private void trySpawn(final ServerLevel level) {
         final io.papermc.paper.threadedregions.RegionizedWorldData worldData = level.getCurrentWorldData(); // Folia - region threading
         Vec3 spawnPos = this.findRandomSpawnPos(level, new BlockPos(worldData.villageSiegeState.spawnX, worldData.villageSiegeState.spawnY, worldData.villageSiegeState.spawnZ)); // Folia - region threading
-        if (spawnPos != null) {
+        if (spawnPos != null && ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(level, spawnPos)) { // okocraft - region threading
             Zombie zombie;
             try {
                 zombie = new Zombie(level);
