--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -978,6 +_,7 @@
     // Paper end - optimise random ticking
 
     public void tickChunk(LevelChunk chunk, int randomTickSpeed) {
+        if (chunk == null) return; // okocraft - Prevent NPE on iterating chunks
         final io.papermc.paper.threadedregions.util.SimpleThreadLocalRandomSource simpleRandom = this.simpleRandom; // Paper - optimise random ticking // Folia - region threading
         ChunkPos pos = chunk.getPos();
         int minBlockX = pos.getMinBlockX();
