--- a/net/minecraft/server/level/ServerPlayer.java
+++ b/net/minecraft/server/level/ServerPlayer.java
@@ -1572,7 +_,16 @@
                         );
                     }
                     // now the respawn logic is complete
-
+                    // okocraft start - Call PlayerRespawnEvent
+                    new org.bukkit.event.player.PlayerRespawnEvent(
+                            this.getBukkitEntity(),
+                            this.getBukkitEntity().getLocation(),
+                            false,
+                            usedRespawnAnchor[0],
+                            false,
+                            reason
+                    );
+                    // okocraft end - Call PlayerRespawnEvent
                     // last, call the function callback
                     if (respawnComplete != null) {
                         respawnComplete.accept(ServerPlayer.this);
@@ -1753,6 +_,7 @@
             this.lastSentFood = -1;
 
             this.triggerDimensionChangeTriggers(originWorld);
+            new org.bukkit.event.player.PlayerChangedWorldEvent(this.getBukkitEntity(), originWorld.getWorld()).callEvent(); // okocraft - Call PlayerChangedWorldEvent
 
             // finished
 
