From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Siroshun09 <siro.shun8@gmail.com>
Date: Wed, 15 Mar 2023 07:49:23 +0900
Subject: [PATCH] okocraft customize server


diff --git a/build.gradle.kts b/build.gradle.kts
index 215366d7cc3720805d2dfde0662372ecede8383c..4d803ed5f42b38945e49566e519ddd1b6ed9fb53 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -8,8 +8,8 @@ plugins {
 
 dependencies {
     // Folia start
-    implementation(project(":folia-api"))
-    implementation("io.papermc.paper:paper-mojangapi:1.19.3-R0.1-SNAPSHOT") {
+    implementation(project(":okocraft-api")) // okocraft
+    implementation("io.papermc.paper:paper-mojangapi:1.20.1-R0.1-SNAPSHOT") { // okocraft
         exclude("io.papermc.paper", "paper-api")
     }
     // Folia end
diff --git a/src/main/java/net/okocraft/paper/OkocraftVersionFetcher.java b/src/main/java/net/okocraft/paper/OkocraftVersionFetcher.java
new file mode 100644
index 0000000000000000000000000000000000000000..e6de952985b5e4f2ff2a64df948fed968f7a73a5
--- /dev/null
+++ b/src/main/java/net/okocraft/paper/OkocraftVersionFetcher.java
@@ -0,0 +1,55 @@
+package net.okocraft.paper;
+
+import com.destroystokyo.paper.VersionHistoryManager;
+import com.destroystokyo.paper.util.VersionFetcher;
+import net.kyori.adventure.text.Component;
+import net.kyori.adventure.text.TextComponent;
+import net.kyori.adventure.text.event.ClickEvent;
+import net.kyori.adventure.text.format.NamedTextColor;
+import net.kyori.adventure.text.format.TextDecoration;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import static net.kyori.adventure.text.Component.newline;
+import static net.kyori.adventure.text.Component.text;
+
+public class OkocraftVersionFetcher implements VersionFetcher {
+
+    private static final String OKOCRAFT_PAPER_REPO = "https://github.com/okocraft/Forked-Paper";
+
+    @Override
+    public long getCacheTime() {
+        return Long.MAX_VALUE; // prevent re-fetching
+    }
+
+    @Override
+    public @NotNull Component getVersionMessage(final @NotNull String serverVersion) {
+        TextComponent.Builder builder =
+                text().append(text("This server software is maintained by OKOCRAFT."))
+                        .append(newline())
+                        .append(text("Repository: "))
+                        .append(text(OKOCRAFT_PAPER_REPO)
+                                .hoverEvent(text("Click to open", NamedTextColor.WHITE))
+                                .clickEvent(ClickEvent.openUrl(OKOCRAFT_PAPER_REPO))
+                        );
+
+
+        final @Nullable Component history = this.getHistory();
+
+        return history != null ? builder.append(newline()).append(history).build() : builder.build();
+    }
+
+    private @Nullable Component getHistory() {
+        final VersionHistoryManager.VersionData data = VersionHistoryManager.INSTANCE.getVersionData();
+        if (data == null) {
+            return null;
+        }
+
+        final String oldVersion = data.getOldVersion();
+        if (oldVersion == null) {
+            return null;
+        }
+
+        return text("Previous version: " + oldVersion, NamedTextColor.GRAY, TextDecoration.ITALIC);
+    }
+}
diff --git a/src/main/java/org/bukkit/craftbukkit/util/CraftMagicNumbers.java b/src/main/java/org/bukkit/craftbukkit/util/CraftMagicNumbers.java
index 61e40979a68017eb730f14a1a0978402f44c1865..e0b7c8668888e942e54219fe3f3e0906be47cd25 100644
--- a/src/main/java/org/bukkit/craftbukkit/util/CraftMagicNumbers.java
+++ b/src/main/java/org/bukkit/craftbukkit/util/CraftMagicNumbers.java
@@ -476,7 +476,7 @@ public final class CraftMagicNumbers implements UnsafeValues {
 
     @Override
     public com.destroystokyo.paper.util.VersionFetcher getVersionFetcher() {
-        return new com.destroystokyo.paper.PaperVersionFetcher();
+        return new net.okocraft.paper.OkocraftVersionFetcher(); // okocraft
     }
 
     @Override
diff --git a/src/main/java/org/bukkit/craftbukkit/util/Versioning.java b/src/main/java/org/bukkit/craftbukkit/util/Versioning.java
index e9b6ca3aa25e140467ae866d572483050ea3fa0e..b2822fee256233d555e6588ba75aa9ac94688a13 100644
--- a/src/main/java/org/bukkit/craftbukkit/util/Versioning.java
+++ b/src/main/java/org/bukkit/craftbukkit/util/Versioning.java
@@ -11,7 +11,7 @@ public final class Versioning {
     public static String getBukkitVersion() {
         String result = "Unknown-Version";
 
-        InputStream stream = Bukkit.class.getClassLoader().getResourceAsStream("META-INF/maven/dev.folia/folia-api/pom.properties"); // Folia
+        InputStream stream = Bukkit.class.getClassLoader().getResourceAsStream("META-INF/maven/net.okocraft.paper/okocraft-api/pom.properties"); // okocraft
         Properties properties = new Properties();
 
         if (stream != null) {
