From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Siroshun09 <siro.shun8@gmail.com>
Date: Sun, 16 Jul 2023 20:55:50 +0900
Subject: [PATCH] Fix not playing travel sound when changing the dimension by
 portals


diff --git a/src/main/java/net/minecraft/server/level/ServerPlayer.java b/src/main/java/net/minecraft/server/level/ServerPlayer.java
index bd9edb8271e3111431d45eb908d8336e0e7e23bc..79009dfa6edafb1e5ee0d475a1a02e7e8bca4a55 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
@@ -1564,6 +1564,7 @@ public class ServerPlayer extends Player {
                 this.connection.send(new ClientboundUpdateMobEffectPacket(this.getId(), mobEffect));
             }
 
+            this.connection.send(new ClientboundLevelEventPacket(net.minecraft.world.level.block.LevelEvent.SOUND_PORTAL_TRAVEL, BlockPos.ZERO, 0, false)); // okocraft - Fix not playing travel sound when changing the dimension by portals
             this.triggerDimensionChangeTriggers(originWorld);
 
             // finished
