From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Siroshun09 <siro.shun8@gmail.com>
Date: Sun, 16 Jul 2023 20:55:50 +0900
Subject: [PATCH] Fix not playing travel sound when changing the dimension by
 portals


diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index eade31ea668c171dceafee958e265041a439c54b..eaa42e9b4caf656fffa154e16a5e12e1b0c4fd8c 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -4091,7 +4091,7 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource, S
             return false;
         }
 
-        return this.portalToAsync(destination, false, PortalType.END, null);
+        return this.portalToAsync(destination, false, PortalType.END, net.okocraft.paper.PortalLevelEventSender.INSTANCE); // okocraft - Fix not playing travel sound when changing the dimension by portals
     }
 
     public boolean netherPortalLogicAsync() {
@@ -4103,7 +4103,7 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource, S
             return false;
         }
 
-        return this.portalToAsync(destination, false, PortalType.NETHER, null);
+        return this.portalToAsync(destination, false, PortalType.NETHER, net.okocraft.paper.PortalLevelEventSender.INSTANCE); // okocraft - Fix not playing travel sound when changing the dimension by portals
     }
 
     private static final java.util.concurrent.atomic.AtomicLong CREATE_PORTAL_DOUBLE_CHECK = new java.util.concurrent.atomic.AtomicLong();
diff --git a/src/main/java/net/okocraft/paper/PortalLevelEventSender.java b/src/main/java/net/okocraft/paper/PortalLevelEventSender.java
new file mode 100644
index 0000000000000000000000000000000000000000..e38cda2801f4bd7f437719afbecb63e837d8427e
--- /dev/null
+++ b/src/main/java/net/okocraft/paper/PortalLevelEventSender.java
@@ -0,0 +1,24 @@
+package net.okocraft.paper;
+
+import net.minecraft.core.BlockPos;
+import net.minecraft.network.protocol.game.ClientboundLevelEventPacket;
+import net.minecraft.server.level.ServerPlayer;
+import net.minecraft.world.entity.Entity;
+import net.minecraft.world.level.block.LevelEvent;
+
+import java.util.function.Consumer;
+
+public final class PortalLevelEventSender implements Consumer<Entity> {
+
+    public static final PortalLevelEventSender INSTANCE = new PortalLevelEventSender();
+
+    private PortalLevelEventSender() {
+    }
+
+    @Override
+    public void accept(Entity entity) {
+        if (entity instanceof ServerPlayer serverPlayer) {
+            serverPlayer.connection.send(new ClientboundLevelEventPacket(LevelEvent.SOUND_PORTAL_TRAVEL, BlockPos.ZERO, 0, false));
+        }
+    }
+}
