From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Siroshun09 <siro.shun8@gmail.com>
Date: Sun, 16 Jul 2023 20:55:50 +0900
Subject: [PATCH] Fix not playing travel sound when changing the dimension by
 portals


diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index c86f00db73d222de4a6dfb97efca8ff555e7bd3c..834b96e29555b9cef653a82c37307d6297c9dab9 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -3977,6 +3977,13 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
         this.processPortalCooldown();
         return false;
     }
+    // okocraft start - Fix not playing travel sound when changing the dimension by portals
+    private static final java.util.function.Consumer<Entity> SEND_TRAVEL_EVENT = entity -> {
+        if (entity instanceof ServerPlayer player) {
+            player.connection.send(new net.minecraft.network.protocol.game.ClientboundLevelEventPacket(1032, BlockPos.ZERO, 0, false));
+        }
+    };
+    // okocraft end
 
     public boolean endPortalLogicAsync() {
         io.papermc.paper.util.TickThread.ensureTickThread(this, "Cannot portal entity async");
@@ -4007,7 +4014,7 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
             return false;
         }
 
-        return this.portalToAsync(destination, false, PortalType.NETHER, null);
+        return this.portalToAsync(destination, false, PortalType.NETHER, SEND_TRAVEL_EVENT); // okocraft - Fix not playing travel sound when changing the dimension by portals
         } // okocraft
     }
 
