From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Siroshun09 <siro.shun8@gmail.com>
Date: Sun, 16 Jul 2023 20:54:11 +0900
Subject: [PATCH] Fix not playing travel sound when changing the dimension by
 portals


diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index c7ec0e148b775e98cb2a41289be1afff41cb25bb..899490bd2d5644b8f62da1ad64c04447e318facd 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -57,7 +57,6 @@ import net.minecraft.network.syncher.EntityDataSerializers;
 import net.minecraft.network.syncher.SynchedEntityData;
 import net.minecraft.resources.ResourceKey;
 import net.minecraft.resources.ResourceLocation;
-import io.papermc.paper.util.MCUtil;
 import net.minecraft.server.MinecraftServer;
 import net.minecraft.server.level.ServerLevel;
 import net.minecraft.server.level.ServerPlayer;
@@ -3976,6 +3975,13 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
         this.processPortalCooldown();
         return false;
     }
+    // okocraft start - Fix not playing travel sound when changing the dimension by portals
+    private static final java.util.function.Consumer<Entity> SEND_TRAVEL_EVENT = entity -> {
+        if (entity instanceof ServerPlayer player) {
+            player.connection.send(new net.minecraft.network.protocol.game.ClientboundLevelEventPacket(1032, BlockPos.ZERO, 0, false));
+        }
+    };
+    // okocraft end
 
     public boolean endPortalLogicAsync() {
         io.papermc.paper.util.TickThread.ensureTickThread(this, "Cannot portal entity async");
@@ -4006,7 +4012,7 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
             return false;
         }
 
-        return this.portalToAsync(destination, false, PortalType.NETHER, null);
+        return this.portalToAsync(destination, false, PortalType.NETHER, SEND_TRAVEL_EVENT); // okocraft - Fix not playing travel sound when changing the dimension by portals
         } // okocraft
     }
 
