From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Siroshun09 <siro.shun8@gmail.com>
Date: Tue, 18 Jul 2023 06:20:25 +0900
Subject: [PATCH] Call spawn/respawn events that cannot modify the location


diff --git a/src/main/java/net/minecraft/server/level/ServerPlayer.java b/src/main/java/net/minecraft/server/level/ServerPlayer.java
index 500b05ce2cda7b2c0e5332e2921d55a3101fc451..1660d58bffeb0cb4e485ec184b5d35a5f4ef11e0 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
@@ -1294,10 +1294,15 @@ public class ServerPlayer extends Player {
 
         this.respawn((player) -> {
             CriteriaTriggers.CHANGED_DIMENSION.trigger(player, Level.END, Level.OVERWORLD);
+            new org.bukkit.event.player.PlayerRespawnEvent(player.getBukkitEntity(), player.getBukkitEntity().getLocation(), false, false, org.bukkit.event.player.PlayerRespawnEvent.RespawnReason.END_PORTAL, com.google.common.collect.ImmutableSet.<org.bukkit.event.player.PlayerRespawnEvent.RespawnFlag>builder().add(org.bukkit.event.player.PlayerRespawnEvent.RespawnFlag.END_PORTAL)).callEvent(); // okocraft - call PlayerRespawnEvent
         }, true);
     }
 
     public void respawn(java.util.function.Consumer<ServerPlayer> respawnComplete) {
+        // okocraft start - call PlayerRespawnEvent
+        java.util.function.Consumer<ServerPlayer> calling = player -> new org.bukkit.event.player.PlayerRespawnEvent(player.getBukkitEntity(), player.getBukkitEntity().getLocation(), false, false, org.bukkit.event.player.PlayerRespawnEvent.RespawnReason.DEATH, com.google.common.collect.ImmutableSet.builder()).callEvent(); // okocraft - call PlayerRespawnEvent
+        respawnComplete = respawnComplete != null ? respawnComplete.andThen(calling) : calling;
+        // okocraft end
         this.respawn(respawnComplete, false);
     }
 
diff --git a/src/main/java/net/minecraft/server/players/PlayerList.java b/src/main/java/net/minecraft/server/players/PlayerList.java
index 7bfbcda5cf086c3c50f8cb202ceaf685a899141c..361e347f651388c5281b106b45e3e882347df357 100644
--- a/src/main/java/net/minecraft/server/players/PlayerList.java
+++ b/src/main/java/net/minecraft/server/players/PlayerList.java
@@ -295,6 +295,9 @@ public abstract class PlayerList {
             worldserver1 = worldserver;
         }
 
+        // okocraft start - Call PlayerSpawnLocationEvent that cannot modify spawn location
+        this.cserver.getPluginManager().callEvent(new org.spigotmc.event.player.PlayerSpawnLocationEvent(player.getBukkitEntity(), worldserver1.getWorld().getSpawnLocation()));
+        // okocraft end
         // Paper start
         if (nbttagcompound == null) {
             player.spawnReason = org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.DEFAULT; // set Player SpawnReason to DEFAULT on first login
