From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Siroshun09 <siro.shun8@gmail.com>
Date: Tue, 18 Jul 2023 06:20:25 +0900
Subject: [PATCH] Call spawn/respawn events that cannot modify the location


diff --git a/src/main/java/net/minecraft/server/level/ServerPlayer.java b/src/main/java/net/minecraft/server/level/ServerPlayer.java
index 42ff3d94ee451ced69c1059f0dbd45085489a60d..76beb06c251a151487ded9adfb061b79838b38a7 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
@@ -1295,10 +1295,15 @@ public class ServerPlayer extends Player {
 
         this.respawn((player) -> {
             CriteriaTriggers.CHANGED_DIMENSION.trigger(player, Level.END, Level.OVERWORLD);
+            new org.bukkit.event.player.PlayerRespawnEvent(player.getBukkitEntity(), player.getBukkitEntity().getLocation(), false, false, org.bukkit.event.player.PlayerRespawnEvent.RespawnReason.END_PORTAL, com.google.common.collect.ImmutableSet.<org.bukkit.event.player.PlayerRespawnEvent.RespawnFlag>builder().add(org.bukkit.event.player.PlayerRespawnEvent.RespawnFlag.END_PORTAL)).callEvent(); // okocraft - call PlayerRespawnEvent
         }, true);
     }
 
     public void respawn(java.util.function.Consumer<ServerPlayer> respawnComplete) {
+        // okocraft start - call PlayerRespawnEvent
+        java.util.function.Consumer<ServerPlayer> calling = player -> new org.bukkit.event.player.PlayerRespawnEvent(player.getBukkitEntity(), player.getBukkitEntity().getLocation(), false, false, org.bukkit.event.player.PlayerRespawnEvent.RespawnReason.DEATH, com.google.common.collect.ImmutableSet.builder()).callEvent(); // okocraft - call PlayerRespawnEvent
+        respawnComplete = respawnComplete != null ? respawnComplete.andThen(calling) : calling;
+        // okocraft end
         this.respawn(respawnComplete, false);
     }
 
diff --git a/src/main/java/net/minecraft/server/players/PlayerList.java b/src/main/java/net/minecraft/server/players/PlayerList.java
index 9da89e8e9569666227081ebc6cb1f0348e1a183d..b19449a9d9b6031f545777629754a78002ab7edf 100644
--- a/src/main/java/net/minecraft/server/players/PlayerList.java
+++ b/src/main/java/net/minecraft/server/players/PlayerList.java
@@ -295,6 +295,9 @@ public abstract class PlayerList {
             worldserver1 = worldserver;
         }
 
+        // okocraft start - Call PlayerSpawnLocationEvent that cannot modify spawn location
+        this.cserver.getPluginManager().callEvent(new org.spigotmc.event.player.PlayerSpawnLocationEvent(player.getBukkitEntity(), worldserver1.getWorld().getSpawnLocation()));
+        // okocraft end
         // Paper start
         if (nbttagcompound == null) {
             player.spawnReason = org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.DEFAULT; // set Player SpawnReason to DEFAULT on first login
