From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Siroshun09 <siro.shun8@gmail.com>
Date: Fri, 22 Mar 2024 19:53:24 +0900
Subject: [PATCH] Checks if the location is owned by current region before
 accessing poi


diff --git a/src/main/java/net/minecraft/world/entity/ai/behavior/YieldJobSite.java b/src/main/java/net/minecraft/world/entity/ai/behavior/YieldJobSite.java
index d1a9b62d3304916275dd6b4c4e783cf1563b5e21..7e55610664d1e34a0bdb02cd67890d305a24e969 100644
--- a/src/main/java/net/minecraft/world/entity/ai/behavior/YieldJobSite.java
+++ b/src/main/java/net/minecraft/world/entity/ai/behavior/YieldJobSite.java
@@ -32,6 +32,7 @@ public class YieldJobSite {
                                     return false;
                                 } else if (entity.getVillagerData().getProfession() != VillagerProfession.NONE) {
                                     return false;
+                                } else if (!ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(world, entity.position())) { return false; // okocraft - region threading
                                 } else {
                                     BlockPos blockPos = context.<GlobalPos>get(potentialJobSite).pos();
                                     Optional<Holder<PoiType>> optional = world.getPoiManager().getType(blockPos);
diff --git a/src/main/java/net/minecraft/world/entity/raid/Raids.java b/src/main/java/net/minecraft/world/entity/raid/Raids.java
index f4f9670e7aa81e3f60b34f0bc7313784573c74b8..b4702eb8c4595e6da53292041b8adc26e493779a 100644
--- a/src/main/java/net/minecraft/world/entity/raid/Raids.java
+++ b/src/main/java/net/minecraft/world/entity/raid/Raids.java
@@ -105,7 +105,7 @@ public class Raids extends SavedData {
         } else {
             DimensionType dimensionmanager = player.level().dimensionType();
 
-            if (!dimensionmanager.hasRaids() || !ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(this.level, player.chunkPosition().x, player.chunkPosition().z, 8)) { // Folia - region threading
+            if (!dimensionmanager.hasRaids() || !ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(this.level, player.chunkPosition().x, player.chunkPosition().z, 10)) { // Folia - region threading // okocraft - fix range
                 return null;
             } else {
                 List<PoiRecord> list = this.level.getPoiManager().getInRange((holder) -> {
