From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Siroshun09 <siro.shun8@gmail.com>
Date: Wed, 19 Jul 2023 18:25:46 +0900
Subject: [PATCH] Call PlayerChangedWorldEvent


diff --git a/src/main/java/net/minecraft/server/level/ServerPlayer.java b/src/main/java/net/minecraft/server/level/ServerPlayer.java
index 46ce4d20c49c362d92bd2454ebe4577bb5fb79a2..7225dab8f8e4dfa47b290a01cc1342eb12c4e3b9 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
@@ -1570,6 +1570,7 @@ public class ServerPlayer extends Player {
             }
 
             this.triggerDimensionChangeTriggers(originWorld);
+            new org.bukkit.event.player.PlayerChangedWorldEvent(this.getBukkitEntity(), originWorld.getWorld()).callEvent(); // okocraft - Call PlayerChangedWorldEvent
 
             // finished
 
