From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Siroshun09 <siro.shun8@gmail.com>
Date: Fri, 4 Aug 2023 11:50:45 +0900
Subject: [PATCH] Disable advancement/recipe warnings


diff --git a/src/main/java/net/minecraft/server/PlayerAdvancements.java b/src/main/java/net/minecraft/server/PlayerAdvancements.java
index 8e2eb7b61421ceb063654826941f1a81f6f50bdf..edc4b0849aecf653718995efe5175a6d24f7b51b 100644
--- a/src/main/java/net/minecraft/server/PlayerAdvancements.java
+++ b/src/main/java/net/minecraft/server/PlayerAdvancements.java
@@ -198,7 +198,7 @@ public class PlayerAdvancements {
             AdvancementHolder advancementholder = loader.get(minecraftkey);
 
             if (advancementholder == null) {
-                if (!minecraftkey.getNamespace().equals("minecraft")) return; // CraftBukkit
+                if (true || !minecraftkey.getNamespace().equals("minecraft")) return; // CraftBukkit // okocraft - Disable advancement warning
                 PlayerAdvancements.LOGGER.warn("Ignored advancement '{}' in progress file {} - it doesn't exist anymore?", minecraftkey, this.playerSavePath);
             } else {
                 this.startProgress(advancementholder, advancementprogress);
diff --git a/src/main/java/net/minecraft/stats/ServerRecipeBook.java b/src/main/java/net/minecraft/stats/ServerRecipeBook.java
index 5c7484ce2850a2eb698a2183b81134b89b0bbcc7..99a33af94029612769ceac9105c9e60834f30959 100644
--- a/src/main/java/net/minecraft/stats/ServerRecipeBook.java
+++ b/src/main/java/net/minecraft/stats/ServerRecipeBook.java
@@ -157,8 +157,8 @@ public class ServerRecipeBook extends RecipeBook {
 
             try {
                 ResourceKey<Recipe<?>> resourcekey = ResourceKey.create(Registries.RECIPE, ResourceLocation.parse(s));
-
                 if (!validPredicate.test(resourcekey)) {
+                    if (false) // okocraft - Disable recipe warnings
                     ServerRecipeBook.LOGGER.error("Tried to load unrecognized recipe: {} removed now.", resourcekey);
                 } else {
                     handler.accept(resourcekey);
