From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Siroshun09 <siro.shun8@gmail.com>
Date: Fri, 4 Aug 2023 11:50:45 +0900
Subject: [PATCH] Disable advancement/recipe warnings


diff --git a/src/main/java/net/minecraft/server/PlayerAdvancements.java b/src/main/java/net/minecraft/server/PlayerAdvancements.java
index 9fabf9322acd663c4452b562494e74aa42eb19da..c65ec58183089be9e0b6644156b20e6b5f14933f 100644
--- a/src/main/java/net/minecraft/server/PlayerAdvancements.java
+++ b/src/main/java/net/minecraft/server/PlayerAdvancements.java
@@ -198,7 +198,7 @@ public class PlayerAdvancements {
             AdvancementHolder advancementholder = loader.get(minecraftkey);
 
             if (advancementholder == null) {
-                if (!minecraftkey.getNamespace().equals("minecraft")) return; // CraftBukkit
+                if (true || !minecraftkey.getNamespace().equals("minecraft")) return; // CraftBukkit // okocraft - Disable advancement warning
                 PlayerAdvancements.LOGGER.warn("Ignored advancement '{}' in progress file {} - it doesn't exist anymore?", minecraftkey, this.playerSavePath);
             } else {
                 this.startProgress(advancementholder, advancementprogress);
diff --git a/src/main/java/net/minecraft/stats/ServerRecipeBook.java b/src/main/java/net/minecraft/stats/ServerRecipeBook.java
index 8a2584deccbf19de8a18fcfcb9c65318b60afe57..869577eb94f631059de9e6190537a666b5b77d22 100644
--- a/src/main/java/net/minecraft/stats/ServerRecipeBook.java
+++ b/src/main/java/net/minecraft/stats/ServerRecipeBook.java
@@ -125,6 +125,7 @@ public class ServerRecipeBook extends RecipeBook {
                 Optional<RecipeHolder<?>> optional = recipeManager.byKey(minecraftkey);
 
                 if (optional.isEmpty()) {
+                    if (false) // okocraft - Disable recipe warnings
                     ServerRecipeBook.LOGGER.error("Tried to load unrecognized recipe: {} removed now.", minecraftkey);
                 } else {
                     handler.accept((RecipeHolder) optional.get());
