From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Siroshun09 <siro.shun8@gmail.com>
Date: Wed, 29 May 2024 20:49:17 +0900
Subject: [PATCH] Fix Firework equality


diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaFirework.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaFirework.java
index 4c09c57c263b18ee2e0156cc4a49ac02a025c651..1368888cb1af0bfec0acc28f3324b8c14110ab9a 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaFirework.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaFirework.java
@@ -242,7 +242,7 @@ class CraftMetaFirework extends CraftMetaItem implements FireworkMeta {
         if (meta instanceof CraftMetaFirework that) {
 
             return (this.hasPower() ? that.hasPower() && this.power == that.power : !that.hasPower())
-                    && (this.effects != null ? that.effects != null && this.effects.equals(that.effects) : that.effects == null); // Paper
+                    && (this.effects != null ? that.effects != null ? this.effects.equals(that.effects) : this.effects.isEmpty() : that.effects == null || that.effects.isEmpty()); // Paper // okocraft
         }
 
         return true;
@@ -260,7 +260,7 @@ class CraftMetaFirework extends CraftMetaItem implements FireworkMeta {
         if (this.hasPower()) {
             hash = 61 * hash + this.power;
         }
-        if (this.effects != null) { // Paper
+        if (this.effects != null && !this.effects.isEmpty()) { // Paper // okocraft
             hash = 61 * hash + 13 * this.effects.hashCode();
         }
         return hash != original ? CraftMetaFirework.class.hashCode() ^ hash : hash;
