From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Siroshun09 <siro.shun8@gmail.com>
Date: Wed, 7 Aug 2024 23:32:04 +0900
Subject: [PATCH] Apply some Paper fixes


diff --git a/src/main/java/net/minecraft/network/protocol/login/ClientboundLoginDisconnectPacket.java b/src/main/java/net/minecraft/network/protocol/login/ClientboundLoginDisconnectPacket.java
index 5d1758086ed4fce5b36a5b31df44ccea42abc590..eeaa40e8121643c6c1d951e76e7361e29210ba48 100644
--- a/src/main/java/net/minecraft/network/protocol/login/ClientboundLoginDisconnectPacket.java
+++ b/src/main/java/net/minecraft/network/protocol/login/ClientboundLoginDisconnectPacket.java
@@ -24,8 +24,9 @@ public class ClientboundLoginDisconnectPacket implements Packet<ClientLoginPacke
     private void write(FriendlyByteBuf buf) {
         // Paper start - Adventure
         // buf.writeUtf(Component.Serializer.toJson(this.reason, RegistryAccess.EMPTY));
-        // In the login phase, buf.adventure$locale field is always null
-        buf.writeJsonWithCodec(net.minecraft.network.chat.ComponentSerialization.localizedCodec(java.util.Locale.US), this.reason, FriendlyByteBuf.MAX_COMPONENT_STRING_LENGTH);
+        // In the login phase, buf.adventure$locale field is most likely null, but plugins may use internals to set it via the channel attribute
+        java.util.Locale bufLocale = buf.adventure$locale;
+        buf.writeJsonWithCodec(net.minecraft.network.chat.ComponentSerialization.localizedCodec(bufLocale == null ? java.util.Locale.US : bufLocale), this.reason, FriendlyByteBuf.MAX_COMPONENT_STRING_LENGTH);
         // Paper end - Adventure
     }
 
diff --git a/src/main/java/net/minecraft/world/level/block/SculkSensorBlock.java b/src/main/java/net/minecraft/world/level/block/SculkSensorBlock.java
index 9e928302817ee8e1cc8c7f3f1b351a0d5f3a7c2c..1db261f63aaf083b4d38f519f77abecbd942e999 100644
--- a/src/main/java/net/minecraft/world/level/block/SculkSensorBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/SculkSensorBlock.java
@@ -218,7 +218,7 @@ public class SculkSensorBlock extends BaseEntityBlock implements SimpleWaterlogg
     }
 
     public static boolean canActivate(BlockState state) {
-        return SculkSensorBlock.getPhase(state) == SculkSensorPhase.INACTIVE;
+        return state.getBlock() instanceof SculkSensorBlock && SculkSensorBlock.getPhase(state) == SculkSensorPhase.INACTIVE; // Paper - Check for a valid type
     }
 
     public static void deactivate(Level world, BlockPos pos, BlockState state) {
