From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Siroshun09 <siro.shun8@gmail.com>
Date: Sun, 19 Feb 2023 07:55:31 +0900
Subject: [PATCH] okocraft Catch and ignore CME that occurred when reloading a
 plugin using ServerUtils


diff --git a/src/main/java/net/minecraft/commands/Commands.java b/src/main/java/net/minecraft/commands/Commands.java
index 4a37225682c3fbc91033da7a0d2800a33d793739..aa22fe53911f20ccfeaf6a3d773b91a1ba325c34 100644
--- a/src/main/java/net/minecraft/commands/Commands.java
+++ b/src/main/java/net/minecraft/commands/Commands.java
@@ -464,7 +464,12 @@ public class Commands {
         // Register Vanilla commands into builtRoot as before
         // Paper start - Async command map building
         COMMAND_SENDING_POOL.execute(() -> {
+            // okocraft start - Catch and ignore CME that occurred when reloading a plugin using ServerUtils
+            try {
                 this.sendAsync(player);
+            } catch (java.util.ConcurrentModificationException ignored) {
+            }
+            // okocraft end
         });
     }
 
