From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Siroshun09 <siro.shun8@gmail.com>
Date: Tue, 20 Sep 2022 14:16:03 +0900
Subject: [PATCH] okocraft add more timings


diff --git a/src/main/java/co/aikar/timings/WorldTimingsHandler.java b/src/main/java/co/aikar/timings/WorldTimingsHandler.java
index fe79c0add4f7cb18d487c5bb9415c40c5b551ea2..af286212e2e6be23ba8eb40aef83dcc7643b1873 100644
--- a/src/main/java/co/aikar/timings/WorldTimingsHandler.java
+++ b/src/main/java/co/aikar/timings/WorldTimingsHandler.java
@@ -67,6 +67,15 @@ public class WorldTimingsHandler {
     public final Timing chunkUnloadPrepareSave;
     public final Timing chunkUnloadPOISerialization;
     public final Timing chunkUnloadDataSave;
+    // okocraft start - add more timings
+    public final Timing entityUpdateVariousThings;
+    public final Timing entityAiStep;
+    public final Timing livingEntityAiStepAi;
+    public final Timing livingEntityAiStepTravel;
+    public final Timing livingEntityAiStepPush;
+
+    public final Timing behaviorAcquirePoiPoiAccess;
+    // okocraft end
 
     public WorldTimingsHandler(Level server) {
         String name = ((PrimaryLevelData) server.getLevelData()).getLevelName() + " - ";
@@ -131,6 +140,17 @@ public class WorldTimingsHandler {
         chunkUnloadPrepareSave = Timings.ofSafe(name + "Chunk unload - Async Save Prepare");
         chunkUnloadPOISerialization = Timings.ofSafe(name + "Chunk unload - POI Data Serialization");
         chunkUnloadDataSave = Timings.ofSafe(name + "Chunk unload - Data Serialization");
+
+        // okocraft start - add more timings
+        entityUpdateVariousThings = Timings.ofSafe(name + "Entity Update Various Things");
+        entityAiStep = Timings.ofSafe(name + "Entity AI Step");
+
+        livingEntityAiStepAi = Timings.ofSafe(name + "LivingEntity AI Step - Ai");
+        livingEntityAiStepTravel = Timings.ofSafe(name + "LivingEntity AI Step - Travel");
+        livingEntityAiStepPush = Timings.ofSafe(name + "LivingEntity AI Step - Push");
+
+        behaviorAcquirePoiPoiAccess = Timings.ofSafe(name + "Behavior AcquirePoi Poi Access");
+        // okocraft end
     }
 
     public static Timing getTickList(ServerLevel worldserver, String timingsType) {
diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index 0a11da6f6242dd016e6016d9f760d01bd43b4bcf..63b220fd73982ac8de1d4e48ee5f5c3deb579c0d 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -816,9 +816,11 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
 
         this.wasInPowderSnow = this.isInPowderSnow;
         this.isInPowderSnow = false;
+        this.level.timings.entityUpdateVariousThings.startTiming(); // okocraft - add more timings
         this.updateInWaterStateAndDoFluidPushing();
         this.updateFluidOnEyes();
         this.updateSwimming();
+        this.level.timings.entityUpdateVariousThings.stopTiming(); // okocraft - add more timings
         if (this.level.isClientSide) {
             this.clearFire();
         } else if (this.remainingFireTicks > 0) {
diff --git a/src/main/java/net/minecraft/world/entity/LivingEntity.java b/src/main/java/net/minecraft/world/entity/LivingEntity.java
index 7c7a61240b9f80a6643cbee17ed9ab8c2e8b9d8a..c6ad62e3e99fedecf611b82ff3b60c162d2ed5cc 100644
--- a/src/main/java/net/minecraft/world/entity/LivingEntity.java
+++ b/src/main/java/net/minecraft/world/entity/LivingEntity.java
@@ -2994,9 +2994,11 @@ public abstract class LivingEntity extends Entity {
             }
         }
 
+        this.level.timings.entityAiStep.startTiming(); // okocraft - add more timing
         if (!this.isRemoved()) {
             this.aiStep();
         }
+        this.level.timings.entityAiStep.stopTiming(); // okocraft - add more timing
 
         double d0 = this.getX() - this.xo;
         double d1 = this.getZ() - this.zo;
@@ -3326,6 +3328,7 @@ public abstract class LivingEntity extends Entity {
         }
 
         this.setDeltaMovement(d4, d5, d6);
+        this.level.timings.livingEntityAiStepAi.startTiming(); // okocraft - add more timings
         this.level.getProfiler().push("ai");
         if (this.isImmobile()) {
             this.jumping = false;
@@ -3336,6 +3339,7 @@ public abstract class LivingEntity extends Entity {
             this.serverAiStep();
             this.level.getProfiler().pop();
         }
+        this.level.timings.livingEntityAiStepAi.stopTiming(); // okocraft - add more timings
 
         this.level.getProfiler().pop();
         this.level.getProfiler().push("jump");
@@ -3365,6 +3369,7 @@ public abstract class LivingEntity extends Entity {
             this.noJumpDelay = 0;
         }
 
+        this.level.timings.livingEntityAiStepTravel.startTiming(); // okocraft - add more timings
         this.level.getProfiler().pop();
         this.level.getProfiler().push("travel");
         this.xxa *= 0.98F;
@@ -3376,6 +3381,7 @@ public abstract class LivingEntity extends Entity {
         this.travel(new Vec3((double) this.xxa, (double) this.yya, (double) this.zza));
         // SpigotTimings.timerEntityAIMove.stopTiming(); // Spigot // Paper
         this.level.getProfiler().pop();
+        this.level.timings.livingEntityAiStepTravel.stopTiming(); // okocraft - add more timings
         this.level.getProfiler().push("freezing");
         boolean flag1 = this.getType().is(EntityTypeTags.FREEZE_HURTS_EXTRA_TYPES);
         int i;
@@ -3397,6 +3403,7 @@ public abstract class LivingEntity extends Entity {
         }
 
         this.level.getProfiler().pop();
+        this.level.timings.livingEntityAiStepPush.startTiming(); // okocraft - add more timings
         this.level.getProfiler().push("push");
         if (this.autoSpinAttackTicks > 0) {
             --this.autoSpinAttackTicks;
@@ -3405,6 +3412,7 @@ public abstract class LivingEntity extends Entity {
 
         this.pushEntities();
         this.level.getProfiler().pop();
+        this.level.timings.livingEntityAiStepPush.stopTiming(); // okocraft - add more timings
         // Paper start
         if (((ServerLevel) this.level).hasEntityMoveEvent && !(this instanceof net.minecraft.world.entity.player.Player)) {
             if (this.xo != getX() || this.yo != this.getY() || this.zo != this.getZ() || this.yRotO != this.getYRot() || this.xRotO != this.getXRot()) {
diff --git a/src/main/java/net/minecraft/world/entity/ai/behavior/AcquirePoi.java b/src/main/java/net/minecraft/world/entity/ai/behavior/AcquirePoi.java
index e46a0e5a080344b5433688994e5aa138d57243b6..a2bc6391827a9d2d9957cae4a93e63c8ba290977 100644
--- a/src/main/java/net/minecraft/world/entity/ai/behavior/AcquirePoi.java
+++ b/src/main/java/net/minecraft/world/entity/ai/behavior/AcquirePoi.java
@@ -88,11 +88,13 @@ public class AcquirePoi extends Behavior<PathfinderMob> {
                 return true;
             }
         };
+        entity.level.timings.behaviorAcquirePoiPoiAccess.startTiming(); // okocraft - add more timings
         // Paper start - optimise POI access
         java.util.List<Pair<Holder<PoiType>, BlockPos>> poiposes = new java.util.ArrayList<>();
         io.papermc.paper.util.PoiAccess.findNearestPoiPositions(poiManager, this.poiType, predicate, entity.blockPosition(), 48, 48*48, PoiManager.Occupancy.HAS_SPACE, false, 5, poiposes);
         Set<Pair<Holder<PoiType>, BlockPos>> set = new java.util.HashSet<>(poiposes);
         // Paper end - optimise POI access
+        entity.level.timings.behaviorAcquirePoiPoiAccess.stopTiming(); // okocraft - add more timings
         Path path = findPathToPois(entity, set);
         if (path != null && path.canReach()) {
             BlockPos blockPos = path.getTarget();
