From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Siroshun09 <siro.shun8@gmail.com>
Date: Sat, 3 Sep 2022 12:05:40 +0900
Subject: [PATCH] okocraft reduce bat spawning


diff --git a/src/main/java/net/minecraft/world/entity/ambient/Bat.java b/src/main/java/net/minecraft/world/entity/ambient/Bat.java
index 50d4595b81f24949011c7565c5e3fc8c26c86019..7e465901c61152f64504cfc463d87b99e1d23272 100644
--- a/src/main/java/net/minecraft/world/entity/ambient/Bat.java
+++ b/src/main/java/net/minecraft/world/entity/ambient/Bat.java
@@ -237,7 +237,7 @@ public class Bat extends AmbientCreature {
     }
 
     public static boolean checkBatSpawnRules(EntityType<Bat> type, LevelAccessor world, MobSpawnType spawnReason, BlockPos pos, RandomSource random) {
-        if (pos.getY() >= world.getSeaLevel()) {
+        if (pos.getY() >= world.getSeaLevel() || random.nextInt(10) != 0) { // okocraft - reduce bat spawning
             return false;
         } else {
             int i = world.getMaxLocalRawBrightness(pos);
