From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Paul Sauve <paul@technove.co>
Date: Mon, 26 Apr 2021 11:34:29 -0500
Subject: [PATCH] Disable builtin method profiler


diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index fb82bb52f219e7683fe1d3c0fb3acbe2251de8d4..ceb6a1c5d3a4800c20860e91c66e790aeeac4530 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -2260,7 +2260,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
     }
 
     public ProfilerFiller getProfiler() {
-        return this.profiler;
+        return net.minecraft.util.profiling.InactiveProfiler.INSTANCE; // Pufferfish
     }
 
     public abstract boolean isSingleplayerOwner(GameProfile profile);
diff --git a/src/main/java/net/minecraft/world/level/Level.java b/src/main/java/net/minecraft/world/level/Level.java
index 3c6ed7a6f1b63083c32d3b4968b668ff85895732..ec2d462184c7c9c4a3714bed34610982afd416de 100644
--- a/src/main/java/net/minecraft/world/level/Level.java
+++ b/src/main/java/net/minecraft/world/level/Level.java
@@ -1462,7 +1462,7 @@ public abstract class Level implements LevelAccessor, AutoCloseable {
     }
 
     public ProfilerFiller getProfiler() {
-        return (ProfilerFiller) this.profiler.get();
+        return (ProfilerFiller) net.minecraft.util.profiling.InactiveProfiler.INSTANCE; // Pufferfish
     }
 
     public Supplier<ProfilerFiller> getProfilerSupplier() {
