From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Siroshun09 <siro.shun8@gmail.com>
Date: Sun, 1 Oct 2023 16:28:36 +0900
Subject: [PATCH] Fix NPE when loading items that created in old version


diff --git a/src/main/java/ca/spottedleaf/dataconverter/minecraft/versions/V3568.java b/src/main/java/ca/spottedleaf/dataconverter/minecraft/versions/V3568.java
index 98272d556a09e9f035005c6783ebb229c2c7157d..60e4eaabd420a27e3055a2c2a52d3c1230fbe919 100644
--- a/src/main/java/ca/spottedleaf/dataconverter/minecraft/versions/V3568.java
+++ b/src/main/java/ca/spottedleaf/dataconverter/minecraft/versions/V3568.java
@@ -116,6 +116,7 @@ public final class V3568 {
     }
 
     private static void convertMobEffectList(final MapType<String> data, final String oldPath, final String newPath) {
+        if (data == null) return; // okocraft - Fix NPE when loading items that created in old version
         final ListType effects = data.getList(oldPath, ObjectType.MAP);
         if (effects == null) {
             return;
@@ -202,6 +203,7 @@ public final class V3568 {
                 final MapType<String> tag = root.getMap("tag");
 
                 if ("minecraft:suspicious_stew".equals(id)) {
+                    if (tag == null) return null; // okocraft - Fix NPE when loading items that created in old version
                     RenameHelper.renameSingle(tag, "Effects", "effects");
 
                     final ListType effects = tag.getList("effects", ObjectType.MAP);
