From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Siroshun09 <siro.shun8@gmail.com>
Date: Tue, 20 Sep 2022 14:16:03 +0900
Subject: [PATCH] okocraft add more timings


diff --git a/src/main/java/co/aikar/timings/WorldTimingsHandler.java b/src/main/java/co/aikar/timings/WorldTimingsHandler.java
index 22687667ec69a954261e55e59261286ac1b8b8cd..c847d7eb9c238ae038e85e07ef91a557081a60c3 100644
--- a/src/main/java/co/aikar/timings/WorldTimingsHandler.java
+++ b/src/main/java/co/aikar/timings/WorldTimingsHandler.java
@@ -68,6 +68,26 @@ public class WorldTimingsHandler {
     public final Timing chunkUnloadPrepareSave;
     public final Timing chunkUnloadPOISerialization;
     public final Timing chunkUnloadDataSave;
+    // okocraft start - add more timings
+    public final Timing entityUpdateVariousThings;
+    public final Timing entityAiStep;
+    public final Timing livingEntityAiStepAi;
+    public final Timing livingEntityAiStepTravel;
+    public final Timing livingEntityAiStepPush;
+
+    public final Timing serverAiStepSensing;
+    public final Timing serverAiStepTargetSelector;
+    public final Timing serverAiStepGoalSelector;
+    public final Timing serverAiStepNavigation;
+    public final Timing serverAiStepMobTick;
+    public final Timing serverAiStepMove;
+    public final Timing serverAiStepLook;
+    public final Timing serverAiStepJump;
+
+    public final Timing villagerBrainTick;
+
+    public final Timing behaviorAcquirePoiPoiAccess;
+    // okocraft end
 
     public WorldTimingsHandler(Level server) {
         String name = ((PrimaryLevelData) server.getLevelData()).getLevelName() + " - ";
@@ -132,6 +152,28 @@ public class WorldTimingsHandler {
         chunkUnloadPrepareSave = Timings.ofSafe(name + "Chunk unload - Async Save Prepare");
         chunkUnloadPOISerialization = Timings.ofSafe(name + "Chunk unload - POI Data Serialization");
         chunkUnloadDataSave = Timings.ofSafe(name + "Chunk unload - Data Serialization");
+
+        // okocraft start - add more timings
+        entityUpdateVariousThings = Timings.ofSafe(name + "Entity Update Various Things");
+        entityAiStep = Timings.ofSafe(name + "Entity AI Step");
+
+        livingEntityAiStepAi = Timings.ofSafe(name + "LivingEntity AI Step - Ai");
+        livingEntityAiStepTravel = Timings.ofSafe(name + "LivingEntity AI Step - Travel");
+        livingEntityAiStepPush = Timings.ofSafe(name + "LivingEntity AI Step - Push");
+
+        serverAiStepSensing = Timings.ofSafe(name + "Server AI Step - Sensing");
+        serverAiStepTargetSelector = Timings.ofSafe(name + "Server AI Step - Target Selector");
+        serverAiStepGoalSelector = Timings.ofSafe(name + "Server AI Step - Goal Selector");
+        serverAiStepNavigation = Timings.ofSafe(name + "Server AI Step - Navigation");
+        serverAiStepMobTick = Timings.ofSafe(name + "Server AI Step - Mob Tick");
+        serverAiStepMove = Timings.ofSafe(name + "Server AI Step - Move");
+        serverAiStepLook = Timings.ofSafe(name + "Server AI Step - Look");
+        serverAiStepJump = Timings.ofSafe(name + "Server AI Step - Jump");
+
+        villagerBrainTick = Timings.ofSafe(name + "Villager Brain Tick");
+
+        behaviorAcquirePoiPoiAccess = Timings.ofSafe(name + "Behavior AcquirePoi Poi Access");
+        // okocraft end
     }
 
     public static Timing getTickList(ServerLevel worldserver, String timingsType) {
diff --git a/src/main/java/net/minecraft/world/entity/LivingEntity.java b/src/main/java/net/minecraft/world/entity/LivingEntity.java
index bef17fe121555450f5f46fb795f30e566561be9c..b894a70cb81a58ebb355064721a18b9fa1683698 100644
--- a/src/main/java/net/minecraft/world/entity/LivingEntity.java
+++ b/src/main/java/net/minecraft/world/entity/LivingEntity.java
@@ -3342,6 +3342,7 @@ public abstract class LivingEntity extends Entity implements Attackable {
             this.serverAiStep();
             this.level.getProfiler().pop();
         }
+        this.level.timings.livingEntityAiStepAi.stopTiming(); // okocraft - add more timings
 
         this.level.getProfiler().pop();
         this.level.getProfiler().push("jump");
diff --git a/src/main/java/net/minecraft/world/entity/Mob.java b/src/main/java/net/minecraft/world/entity/Mob.java
index ee40b2763062bcc1649acec55552f87474ca3a2e..fd3e472b51aad68e6df1209d4458dc85e10a11bd 100644
--- a/src/main/java/net/minecraft/world/entity/Mob.java
+++ b/src/main/java/net/minecraft/world/entity/Mob.java
@@ -903,7 +903,9 @@ public abstract class Mob extends LivingEntity implements Targeting {
         }
         // Paper end
         this.level.getProfiler().push("sensing");
+        this.level.timings.serverAiStepSensing.startTiming();
         this.sensing.tick();
+        this.level.timings.serverAiStepSensing.stopTiming();
         this.level.getProfiler().pop();
         int i = this.level.getServer().getTickCount() + this.getId();
 
diff --git a/src/main/java/net/minecraft/world/entity/npc/Villager.java b/src/main/java/net/minecraft/world/entity/npc/Villager.java
index 5402a084ef5fe0b3cfea897a90cffade1eff5b66..c72e8ee9943ab692f40966185effc36616247fba 100644
--- a/src/main/java/net/minecraft/world/entity/npc/Villager.java
+++ b/src/main/java/net/minecraft/world/entity/npc/Villager.java
@@ -250,12 +250,14 @@ public class Villager extends AbstractVillager implements ReputationEventHandler
     protected void customServerAiStep() { mobTick(false); }
     protected void mobTick(boolean inactive) {
         this.level.getProfiler().push("villagerBrain");
+        this.level.timings.villagerBrainTick.startTiming(); // okocraft - add more timings
         // Pufferfish start
         if (!inactive) {
             if (this.behaviorTick++ % this.activatedPriority == 0) // Pufferfish
             this.getBrain().tick((ServerLevel) this.level, this); // Paper
          }
         // Pufferfish end
+        this.level.timings.villagerBrainTick.stopTiming(); // okocraft - add more timings
         this.level.getProfiler().pop();
         if (this.assignProfessionWhenSpawned) {
             this.assignProfessionWhenSpawned = false;
