From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Siroshun09 <siro.shun8@gmail.com>
Date: Sun, 19 Feb 2023 07:55:31 +0900
Subject: [PATCH] okocraft Catch and ignore CME that occurred when reloading a
 plugin using ServerUtils


diff --git a/src/main/java/net/minecraft/commands/Commands.java b/src/main/java/net/minecraft/commands/Commands.java
index 87cc7562e4a166d078fe11b7f6980497fc0bd33e..77be8d5f0dbcc5a7216ec67a60136279aa5831a0 100644
--- a/src/main/java/net/minecraft/commands/Commands.java
+++ b/src/main/java/net/minecraft/commands/Commands.java
@@ -411,7 +411,12 @@ public class Commands {
         // Register Vanilla commands into builtRoot as before
         // Paper start - Async command map building
         COMMAND_SENDING_POOL.execute(() -> {
+            // okocraft start - Catch and ignore CME that occurred when reloading a plugin using ServerUtils
+            try {
                 this.sendAsync(player);
+            } catch (java.util.ConcurrentModificationException ignored) {
+            }
+            // okocraft end
         });
     }
 
