From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Siroshun09 <siro.shun8@gmail.com>
Date: Sun, 19 Feb 2023 07:55:31 +0900
Subject: [PATCH] okocraft Catch and ignore CME that occurred when reloading a
 plugin using ServerUtils


diff --git a/src/main/java/net/minecraft/commands/Commands.java b/src/main/java/net/minecraft/commands/Commands.java
index e92864ecf32dd984f6f87f7b05341e43af3a2977..c4fd805eaccc6d03c4efd9699371cae7b93efdfb 100644
--- a/src/main/java/net/minecraft/commands/Commands.java
+++ b/src/main/java/net/minecraft/commands/Commands.java
@@ -405,7 +405,12 @@ public class Commands {
         // Register Vanilla commands into builtRoot as before
         // Paper start - Async command map building
         COMMAND_SENDING_POOL.execute(() -> {
+            // okocraft start - Catch and ignore CME that occurred when reloading a plugin using ServerUtils
+            try {
                 this.sendAsync(player);
+            } catch (java.util.ConcurrentModificationException ignored) {
+            }
+            // okocraft end
         });
     }
 
