From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Siroshun09 <siro.shun8@gmail.com>
Date: Wed, 22 Feb 2023 04:05:05 +0900
Subject: [PATCH] okocraft Skip dependency checking to allow bad reloads using
 PluginManager#loadPlugin


diff --git a/src/main/java/io/papermc/paper/plugin/entrypoint/strategy/ModernPluginLoadingStrategy.java b/src/main/java/io/papermc/paper/plugin/entrypoint/strategy/ModernPluginLoadingStrategy.java
index a0363af157131a2b42c00028a260e9c3d3c543de..d45074a6e51bdb6a2252e6f35e2d5b226c981478 100644
--- a/src/main/java/io/papermc/paper/plugin/entrypoint/strategy/ModernPluginLoadingStrategy.java
+++ b/src/main/java/io/papermc/paper/plugin/entrypoint/strategy/ModernPluginLoadingStrategy.java
@@ -18,6 +18,7 @@ import java.util.Map;
 
 @SuppressWarnings("UnstableApiUsage")
 public class ModernPluginLoadingStrategy<T> implements ProviderLoadingStrategy<T> {
+    public static boolean skipDependencyCheck = false; // okocraft - Skip dependency checking to allow bad reloads using PluginManager#loadPlugin
 
     private static final Logger LOGGER = LogUtils.getLogger();
     private final ProviderConfiguration<T> configuration;
@@ -63,6 +64,12 @@ public class ModernPluginLoadingStrategy<T> implements ProviderLoadingStrategy<T
 
         // Validate providers, ensuring all of them have valid dependencies. Removing those who are invalid
         for (PluginProvider<T> provider : pluginProviders) {
+            // okocraft start - Skip dependency checking to allow bad reloads using PluginManager#loadPlugin
+            if (skipDependencyCheck) {
+                validatedProviders.add(provider);
+                continue;
+            }
+            // okocraft end
             PluginMeta configuration = provider.getMeta();
 
             // Populate missing dependencies to capture if there are multiple missing ones.
diff --git a/src/main/java/io/papermc/paper/plugin/manager/PaperPluginInstanceManager.java b/src/main/java/io/papermc/paper/plugin/manager/PaperPluginInstanceManager.java
index c0e896343c22badd97c774c4ed1daa4e274f5d44..fbb8a9827009941cf76ff72d5c5cac7706ee3913 100644
--- a/src/main/java/io/papermc/paper/plugin/manager/PaperPluginInstanceManager.java
+++ b/src/main/java/io/papermc/paper/plugin/manager/PaperPluginInstanceManager.java
@@ -118,9 +118,14 @@ class PaperPluginInstanceManager {
         }
 
         try {
+            io.papermc.paper.plugin.entrypoint.strategy.ModernPluginLoadingStrategy.skipDependencyCheck = true; // okocraft - Skip dependency checking to allow bad reloads using PluginManager#loadPlugin
             runtimePluginEntrypointHandler.enter(Entrypoint.PLUGIN);
         } catch (Throwable e) {
             throw new InvalidPluginException(e);
+        // okocraft start - Skip dependency checking to allow bad reloads using PluginManager#loadPlugin
+        } finally {
+            io.papermc.paper.plugin.entrypoint.strategy.ModernPluginLoadingStrategy.skipDependencyCheck = false;
+        // okocraft end
         }
 
         return runtimePluginEntrypointHandler.getPluginProviderStorage().getSingleLoaded()
