From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Siroshun09 <siro.shun8@gmail.com>
Date: Wed, 9 Aug 2023 19:00:31 +0900
Subject: [PATCH] Run POI modification on correct tick thread


diff --git a/src/main/java/net/minecraft/world/entity/ai/behavior/GoToPotentialJobSite.java b/src/main/java/net/minecraft/world/entity/ai/behavior/GoToPotentialJobSite.java
index aac1df9e2c0652bc6348af0404deba7465f82d42..3e815dd10496f30ef13b06cb27662e437252a507 100644
--- a/src/main/java/net/minecraft/world/entity/ai/behavior/GoToPotentialJobSite.java
+++ b/src/main/java/net/minecraft/world/entity/ai/behavior/GoToPotentialJobSite.java
@@ -48,12 +48,23 @@ public class GoToPotentialJobSite extends Behavior<Villager> {
             BlockPos blockPos = pos.pos();
             ServerLevel serverLevel2 = serverLevel.getServer().getLevel(pos.dimension());
             if (serverLevel2 != null) {
+                Runnable run = () -> { // okocraft - region threading
                 PoiManager poiManager = serverLevel2.getPoiManager();
                 if (poiManager.exists(blockPos, poiType -> true)) {
                     poiManager.release(blockPos);
                 }
 
                 DebugPackets.sendPoiTicketCountPacket(serverLevel, blockPos);
+                // okocraft start - region threading
+                };
+                if (io.papermc.paper.util.TickThread.isTickThreadFor(serverLevel2, blockPos)) {
+                    run.run();
+                } else {
+                    io.papermc.paper.threadedregions.RegionizedServer.getInstance().taskQueue.queueChunkTask(
+                            serverLevel2, blockPos.getX() >> 4, blockPos.getZ() >> 4, run
+                    );
+                }
+                // okocraft end - region threading
             }
         });
         villager.getBrain().eraseMemory(MemoryModuleType.POTENTIAL_JOB_SITE);
