From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Siroshun09 <siro.shun8@gmail.com>
Date: Wed, 30 Aug 2023 20:43:34 +0900
Subject: [PATCH] Ensure that the zombie spawn point is on the current tick
 thread


diff --git a/src/main/java/net/minecraft/world/entity/ai/village/VillageSiege.java b/src/main/java/net/minecraft/world/entity/ai/village/VillageSiege.java
index a2bc957bd643b90595711270ab88382c957e24b6..14901b40a251499c7d3941d527ab1c572bafb821 100644
--- a/src/main/java/net/minecraft/world/entity/ai/village/VillageSiege.java
+++ b/src/main/java/net/minecraft/world/entity/ai/village/VillageSiege.java
@@ -114,7 +114,7 @@ public class VillageSiege implements CustomSpawner {
         io.papermc.paper.threadedregions.RegionizedWorldData worldData = world.getCurrentWorldData(); // Folia - region threading
         Vec3 vec3d = this.findRandomSpawnPos(world, new BlockPos(worldData.villageSiegeState.spawnX, worldData.villageSiegeState.spawnY, worldData.villageSiegeState.spawnZ)); // Folia - region threading
 
-        if (vec3d != null) {
+        if (vec3d != null && ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(world, vec3d)) { // okocraft - region threading
             Zombie entityzombie;
 
             try {
