From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Siroshun09 <siro.shun8@gmail.com>
Date: Wed, 30 Aug 2023 20:43:34 +0900
Subject: [PATCH] Ensure that the zombie spawn point is on the current tick
 thread


diff --git a/src/main/java/net/minecraft/world/entity/ai/village/VillageSiege.java b/src/main/java/net/minecraft/world/entity/ai/village/VillageSiege.java
index 316b5db25f749854052449c1a2a48f20b8ead05b..77846fb36660e9021a9189074ed07c8fc8998603 100644
--- a/src/main/java/net/minecraft/world/entity/ai/village/VillageSiege.java
+++ b/src/main/java/net/minecraft/world/entity/ai/village/VillageSiege.java
@@ -114,7 +114,7 @@ public class VillageSiege implements CustomSpawner {
         io.papermc.paper.threadedregions.RegionizedWorldData worldData = world.getCurrentWorldData(); // Folia - region threading
         Vec3 vec3d = this.findRandomSpawnPos(world, new BlockPos(worldData.villageSiegeState.spawnX, worldData.villageSiegeState.spawnY, worldData.villageSiegeState.spawnZ)); // Folia - region threading
 
-        if (vec3d != null) {
+        if (vec3d != null && io.papermc.paper.util.TickThread.isTickThreadFor(world, vec3d)) { // okocraft - region threading
             Zombie entityzombie;
 
             try {
