From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Siroshun09 <siro.shun8@gmail.com>
Date: Sun, 19 Feb 2023 07:55:31 +0900
Subject: [PATCH] okocraft Catch and ignore CME that occurred when reloading a
 plugin using ServerUtils


diff --git a/src/main/java/net/minecraft/commands/Commands.java b/src/main/java/net/minecraft/commands/Commands.java
index 3435bdeaf723c64103f7c924ea42a4ec78f2ba01..f7f3d28bf454a54cfd52551894cabdf8ac5a570b 100644
--- a/src/main/java/net/minecraft/commands/Commands.java
+++ b/src/main/java/net/minecraft/commands/Commands.java
@@ -419,7 +419,12 @@ public class Commands {
         // Register Vanilla commands into builtRoot as before
         // Paper start - Async command map building
         COMMAND_SENDING_POOL.execute(() -> {
+            // okocraft start - Catch and ignore CME that occurred when reloading a plugin using ServerUtils
+            try {
                 this.sendAsync(player);
+            } catch (java.util.ConcurrentModificationException ignored) {
+            }
+            // okocraft end
         });
     }
 
