--- a/folia-api/build.gradle.kts
+++ b/folia-api/build.gradle.kts
@@ -102,6 +_,7 @@
         java {
             srcDir(generatedDir)
             srcDir(file("../paper-api/src/main/java"))
+            srcDir(file("../folia-api/src/main/java")) // okocraft - Build Changes - add folia-api to source sets
         }
         resources {
             srcDir(file("../paper-api/src/main/resources"))
@@ -194,6 +_,7 @@
 val services = objects.newInstance<Services>()
 
 tasks.withType<Javadoc>().configureEach {
+    enabled = false // okocraft - Build Changes - Javadoc generation fails due to javadoc.io sometimes down. Disable this since this is not needed for us.
     val options = options as StandardJavadocDocletOptions
     options.overview = "../paper-api/src/main/javadoc/overview.html"
     options.use()
